File: movecirclepoint/src/main/java/com/wan/movecirclepoint/CirclePoint.java
package com.wan.movecirclepoint;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutCompat;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

/**
 * Created by xen on 2018/2/7 0007.
 */

public class CirclePoint extends LinearLayout {
    private float size;
    private int point_unselected_color,point_selected_color,unselected_drawble_color,selected_drawble_color;
    private Drawable unsselected_drawble,selected_drawble;
    private RelativeLayout centerLayout;
    private LinearLayout center,pictureLayout;

    private int count;
    private ImageView whitePoint;
    private int mPointMargin;
    private boolean allowCustom;
    private GradientDrawable selected_point,unselected_point,selected_picture,unselected_picture,selected,unselected;
    public CirclePoint(Context context) {
        super(context);
    }

    public CirclePoint(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setOrientation(VERTICAL);
        handlePicture();
        initLayout(context);
        init(context,attrs);

    }

    public void initLayout(Context context){
        centerLayout = new RelativeLayout(context);
        pictureLayout = new LinearLayout(context);
        whitePoint = new ImageView(context);

    }
    public CirclePoint(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }


    public void initImage(int count,Context context){
        whitePoint.setImageDrawable(selected);
        // 设置白点的布局参数
        int pointSize = (int)size;
        //生成一个正方形的RelativeLayout，将ImageView设置为一个固定大小正方形（动态改变ImageView的宽高）
        RelativeLayout.LayoutParams params1 = new RelativeLayout.LayoutParams(pointSize, pointSize);
        whitePoint.setLayoutParams(params1);

        for (int i = 0; i < count; i++) {
            // 设置底部小圆点(灰色)
            ImageView point = new ImageView(context);
            point.setImageDrawable(unselected);
            // 设置灰色点的布局参数
            LayoutParams params2 = new LayoutParams(pointSize, pointSize);
            if (i > 0) {
                params2.leftMargin = (int)size;
            }
            point.setLayoutParams(params2);
            pictureLayout.addView(point);

        }
    }

    public void setPoint(){
        whitePoint.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                // 此时layout布局已经显示出来了，可以获取小圆点之间的距离了
                mPointMargin = pictureLayout.getChildAt(1).getLeft() - pictureLayout.getChildAt(0).getLeft();
                // 将自己移除掉
                whitePoint.getViewTreeObserver().removeOnGlobalLayoutListener(this);
            }
        });
    }
    public void setonPageScrolled(int position,float positionOffset, int positionOffsetPixels){
        int leftMargin = (int) (mPointMargin * (position + positionOffset)); // 页面滑动的时候，动态的获取小圆点的左边距
        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) whitePoint.getLayoutParams();// 获取布局参数，然后设置布局参数
        params.leftMargin = leftMargin; // 修改参数
        whitePoint.setLayoutParams(params); // 重新设置布局参数
    }
    private void init(Context context,AttributeSet attrs){
        TypedArray typedArray = context.obtainStyledAttributes(attrs,R.styleable.CirclePoint);

        size = typedArray.getDimension(R.styleable.CirclePoint_size,8);

        point_selected_color = typedArray.getColor(R.styleable.CirclePoint_point_selected_color,0);
        point_unselected_color = typedArray.getColor(R.styleable.CirclePoint_point_unselected_color,0);
        unselected_drawble_color = typedArray.getColor(R.styleable.CirclePoint_unselected_drawble_color,0);
        selected_drawble_color = typedArray.getColor(R.styleable.CirclePoint_selected_drawble_color,0);

        selected_drawble = typedArray.getDrawable(R.styleable.CirclePoint_selected_drawble);
        unsselected_drawble = typedArray.getDrawable(R.styleable.CirclePoint_unselected_drawble);
        count = typedArray.getInteger(R.styleable.CirclePoint_count,3);
        allowCustom = typedArray.getBoolean(R.styleable.CirclePoint_Allow_custom,false);
        typedArray.recycle();

        setColor();

        centerLayout.setLayoutParams(new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,ViewGroup.LayoutParams.MATCH_PARENT));
        centerLayout.setGravity(Gravity.CENTER_HORIZONTAL);

        pictureLayout.setLayoutParams(new LinearLayoutCompat.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        pictureLayout.setOrientation(HORIZONTAL);

        initImage(count,context);//初始化圆点的图片
        centerLayout.addView(pictureLayout);
        centerLayout.addView(whitePoint);

        center= new LinearLayout(context);
        LayoutParams layoutParams = new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,ViewGroup.LayoutParams.MATCH_PARENT);
        layoutParams.gravity = Gravity.CENTER_HORIZONTAL;
        center.setLayoutParams(layoutParams);
        center.addView(centerLayout);//添加RelativeLayout
        addView(center);

    }
    private void handlePicture(){
        if(Build.VERSION.SDK_INT<21){
            handlePictureBefore();
        }else {
            handlePictureOn();
        }
    }
    private void setColor(){
        selected = new GradientDrawable();
        unselected = new GradientDrawable();

        if (point_selected_color!=0){
            selected_point.setColor(point_selected_color);
        }
        if(point_unselected_color!=0){
            unselected_point.setColor(point_unselected_color);
        }
        if(selected_drawble_color!=0){
            selected_picture.setColor(selected_drawble_color);
        }
        if(unselected_drawble_color!=0){
            unselected_picture.setColor(unselected_drawble_color);
        }

        if (allowCustom){
            selected = selected_picture;
            unselected = unselected_picture;
        }
        selected = selected_point;
        unselected = unselected_point;

    }
    @TargetApi(21)
    private void handlePictureOn(){
            selected_point= (GradientDrawable) getResources().getDrawable(R.drawable.shape_point_selected,null);
            unselected_point = (GradientDrawable)getResources().getDrawable(R.drawable.shape_point_normal,null);
            selected_picture = (GradientDrawable)selected_drawble;
            unselected_picture = (GradientDrawable)unsselected_drawble;

    }
    private void handlePictureBefore(){
        selected_point= (GradientDrawable) getResources().getDrawable(R.drawable.shape_point_selected);
        unselected_point = (GradientDrawable)getResources().getDrawable(R.drawable.shape_point_normal);
        selected_picture = (GradientDrawable)selected_drawble;
        unselected_picture = (GradientDrawable)unsselected_drawble;
    }
}

File: movecirclepoint/src/main/java/com/wan/movecirclepoint/CirclePoint.kt
package com.wan.movecirclepoint

import android.annotation.TargetApi
import android.content.Context
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.util.AttributeSet
import android.view.Gravity
import android.view.ViewGroup
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import androidx.appcompat.widget.LinearLayoutCompat

/**
 * Created by xen on 2018/2/7 0007.
 */
class CirclePoint : LinearLayout {
    private var size = 0f
    private var point_unselected_color = 0
    private var point_selected_color = 0

    private lateinit var selected_point: GradientDrawable
    private lateinit var unselected_point: GradientDrawable


    private val centerLayout by lazy { RelativeLayout(context) }
    private val center by lazy { LinearLayout(context) }
    private val pictureLayout by lazy { LinearLayout(context) }

    private var count = 0
    private val whitePoint by lazy { ImageView(context) }
    private var mPointMargin = 0

    constructor(context: Context?) : super(context) {}
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        orientation = VERTICAL
        handlePicture()

        readStylePro(attrs)
        init(context)
    }


    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    fun initImage(count: Int, context: Context?) {
        whitePoint!!.setImageDrawable(selected_point)
        // 设置白点的布局参数
        val pointSize = size.toInt()
        //生成一个正方形的RelativeLayout，将ImageView设置为一个固定大小正方形（动态改变ImageView的宽高）
        val params1 = RelativeLayout.LayoutParams(pointSize, pointSize)
        whitePoint!!.layoutParams = params1
        for (i in 0 until count) {
            // 设置底部小圆点(灰色)
            val point = ImageView(context)
            point.setImageDrawable(unselected_point)
            // 设置灰色点的布局参数
            val params2 = LayoutParams(pointSize, pointSize)
            if (i > 0) {
                params2.leftMargin = size.toInt()
            }
            point.layoutParams = params2
            pictureLayout!!.addView(point)
        }
    }

    /**
     * 设置圆点数量
     */
    fun setCount(count: Int) {
        pictureLayout.removeAllViews()
        centerLayout.removeAllViews()
        center.removeAllViews()

        removeAllViews()

        this.count = count
        init(context)
    }


    private fun setPoint() {
        whitePoint!!.viewTreeObserver.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                // 此时layout布局已经显示出来了，可以获取小圆点之间的距离了
                mPointMargin = pictureLayout!!.getChildAt(1).left - pictureLayout!!.getChildAt(0).left
                // 将自己移除掉
                whitePoint!!.viewTreeObserver.removeOnGlobalLayoutListener(this)
            }
        })
    }

    fun setonPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        val leftMargin = (mPointMargin * (position + positionOffset)).toInt() // 页面滑动的时候，动态的获取小圆点的左边距
        val params = whitePoint!!.layoutParams as RelativeLayout.LayoutParams // 获取布局参数，然后设置布局参数
        params.leftMargin = leftMargin // 修改参数
        whitePoint!!.layoutParams = params // 重新设置布局参数
    }

    private fun readStylePro(attrs: AttributeSet?) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.CirclePoint)
        size = typedArray.getDimension(R.styleable.CirclePoint_size, 8f)
        point_selected_color = typedArray.getColor(R.styleable.CirclePoint_point_selected_color, 0)
        point_unselected_color = typedArray.getColor(R.styleable.CirclePoint_point_unselected_color, 0)
        count = typedArray.getInteger(R.styleable.CirclePoint_count, 3)
        typedArray.recycle()
    }

    private fun init(context: Context) {
        setColor()

        centerLayout!!.layoutParams = RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        centerLayout!!.gravity = Gravity.CENTER_HORIZONTAL
        pictureLayout!!.layoutParams = LinearLayoutCompat.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        pictureLayout!!.orientation = HORIZONTAL
        initImage(count, context) //初始化圆点的图片

        centerLayout!!.addView(pictureLayout)
        centerLayout!!.addView(whitePoint)

        val layoutParams = LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT)
        layoutParams.gravity = Gravity.CENTER_HORIZONTAL
        center!!.layoutParams = layoutParams
        center!!.addView(centerLayout) //添加RelativeLayout
        addView(center)
        setPoint()
    }

    private fun handlePicture() {
        if (Build.VERSION.SDK_INT < 21) {
            handlePictureBefore()
        } else {
            handlePictureOn()
        }
    }

    private fun setColor() {
        if (point_selected_color != 0) {
            selected_point!!.setColor(point_selected_color)
        }
        if (point_unselected_color != 0) {
            unselected_point!!.setColor(point_unselected_color)
        }
    }

    @TargetApi(21)
    private fun handlePictureOn() {
        selected_point = resources.getDrawable(R.drawable.shape_point_selected, null) as GradientDrawable
        unselected_point = resources.getDrawable(R.drawable.shape_point_normal, null) as GradientDrawable
    }

    private fun handlePictureBefore() {
        selected_point = resources.getDrawable(R.drawable.shape_point_selected) as GradientDrawable
        unselected_point = resources.getDrawable(R.drawable.shape_point_normal) as GradientDrawable
    }
}